src/main/java/caca/TaskList.java

```diff
@@ -0,0 +1,285 @@
+ * <li>Mark task (marks task as done with a "X"): mark taskIndex. e.g. mark 2</li>
+ * <li>Unmark task (marks task as not done and removes "X"): unmark taskIndex. e.g. unmark 2</li>
+ * <li>Delete task (deletes task from list): delete taskIndex. e.g. delete 3</li>
+ * <li>Find task (finds all matching tasks from list): find task. e.g. find book</li>
+ * </ul>
+ * </p>
+ *
+ * @author Carrie Zheng Jiarui
+ * @version CS2103T AY22/23 Semester 1, iP
+ */
+public class TaskList {
+
+    /**
+     * An array to store the list of tasks input by user.
+     */
+    private static List<Task> tasks = new ArrayList<>();
+
+    /**
+     * Constructor for creating a TaskList.
+     *
+     * @param tasks The list of tasks.
+     */
+    public TaskList(ArrayList<Task> tasks) {
+        if (tasks == null) {
+            this.tasks = new ArrayList<>();
+        } else {
+            this.tasks = tasks;
+        }
+    }
+
+    /**
+     * Gets the list of tasks.
+     *
+     * @return The list of tasks.
+     */
+    public List<Task> getTasks() {
+        return tasks;
+    }
+
+    /**
+     * Adds a task to the list.
+     *
+     * @param task Task to be added to the list.
+     */
+    public void addTask(Task task) {
+        tasks.add(task);
+    }
+
+    /**
+     * Checks if task index is valid.
+     *
+     * @param taskIndex Task index entered by user.
+     * @throws InvalidIndexException If task index is invalid, i.e. out of range.
+     */
+    public static void isValid(int taskIndex) throws InvalidIndexException {
+        if (taskIndex <= 0 || taskIndex > tasks.size()) {
+            String message = String.format("OOPS!!! You have entered an invalid task index. "
+                    + "It should be between 1 and %d.", tasks.size());
+            throw new InvalidIndexException(message);
+        }
+    }
+
+    /**
+     * Checks if task description is empty or contains only white spaces.
+     *
+     * @param command User input with 1st element as command type, 2nd element as task description.
+     * @throws EmptyInputException If task description is empty or left blank.
+     */
+    public static void hasDescription(String[] command) throws EmptyInputException {
+        String commandType = command[0];
+        if (command.length == 1 || command[1].isBlank()) {
+            String message = String.format("OOPS!!! The description of %s cannot be empty.",
+                    commandType);
+            throw new EmptyInputException(message);
+        }
+    }
+
+    /**
+     * Adds a ToDo task to user list.
+     *
+     * @param taskDescription Task description entered by user.
+     */
+    public static void addToDo(String taskDescription) {
+        Task taskToAdd = new Todo(taskDescription);
+        tasks.add(taskToAdd);
+
+        System.out.println("Got it. I've added this task:");
+        System.out.println(taskToAdd);
+        System.out.printf("Now you have %d tasks in the list.\n", tasks.size());
```

Comment:
Similarly to my above comment, do you feel you could make use of your Ui class to handle all display messages?

Reply:
Will try to improve on that, thanks! 😄